### ISSUE TITLE
difficult to parse a basic language

### ISSUE THREAD

**Issue by rain-1** [rain-1/racket-peg#67]
https://www.rosettacode.org/wiki/Brace_expansion

```
#lang peg

item <- [^{},]+ / ~'{' choice ~'}';
choice <- x:item+ xs:(~',' item+)* -> (cons x xs);

; > (peg (+ item) "It{{em,alic}iz,erat}e{d,}")
; '("It" ((("em" . "alic") "iz") . "erat") "e")
```

The resulting parse is confusing and we can't change the char class to *. it's too hard to parse this simple language.

**Comment by petrolifero** [comment]
```
#lang peg

braceLanguage <- item+;

item <- braceExpansion / simpleItem;
braceExpansion <- '{' braceLanguage (',' braceLanguage)* ','? '}';

simpleItem <- [^{,}]+;
```


this works on that example

**Comment by petrolifero** [comment]
```
#lang peg

(struct seq (a b) #:transparent);
(struct brace (l) #:transparent);
(struct idt (a) #:transparent);

braceLanguage <- v:item+ -> (let loop ((l v))
                              (if (null? (cdr l))
                                  (car l)
                                  (loop (cons (seq (car l) (cadr l)) (cddr l)))));

item <- braceExpansion / simpleItem;
braceExpansion <- ~'{' l:(braceLanguage (~',' braceLanguage)* ~(','?)) '}' -> l;

simpleItem <- v:[^{,}]+ -> (idt v);
```


with semantic actions

**Comment by petrolifero** [comment]
With semantic actions, we really need to use this `(a . (b . c))` thing? Is not simpler to create structs?

**Comment by petrolifero** [comment]
> ```
> #lang peg
> 
> (struct seq (a b) #:transparent);
> (struct brace (l) #:transparent);
> (struct idt (a) #:transparent);
> 
> braceLanguage <- v:item+ -> (let loop ((l v))
>                               (if (null? (cdr l))
>                                   (car l)
>                                   (loop (cons (seq (car l) (cadr l)) (cddr l)))));
> 
> item <- braceExpansion / simpleItem;
> braceExpansion <- ~'{' l:(braceLanguage (~',' braceLanguage)* ~(','?)) '}' -> l;
> 
> simpleItem <- v:[^{,}]+ -> (idt v);
> ```
> with semantic actions

To me, the answer from this grammar is not confusing.  `(list a b)` represent `{a,b}`, 
`(seq a b)` represent `ab` where a and b are not both simpleItems

**Comment by petrolifero** [comment]
Whats some precise metrics of what you want?

**Comment by rain-1** [comment]
I'm happy with this solution:

```
#lang racket

;; list-monad.rkt

(provide liftm2 sequence
         concatenate string-concatenate)

(define (concatenate lists)
  (apply append lists))

(define (string-concatenate strings)
  (apply string-append strings))

(define (>>= m f)
  (apply append (map f m)))

(define (liftm2 f a b)
  (>>= a (lambda (x) (>>= b (lambda (y) (list (f x y)))))))

(define (sequence lsts)
  (if (null? lsts)
      (list '())
      (liftm2 cons (car lsts) (sequence (cdr lsts)))))
```

```
#lang peg

(require "list-monad.rkt");

text <- str:[^{},]+ -> (list str);
items <- xs:(text / choice)+ -> (map string-concatenate (sequence xs));
choice <- '{' x:items xs:(~',' items)* '}' -> (concatenate (cons x xs));
```

**Comment by petrolifero** [comment]
Cool. Using monads
